### PR TITLE
fix compilation error when boost < 1.68 and libprotobuf < 3.6 

### PR DESCRIPTION
**Pull request by xlqian**
# Issue https://github.com/valhalla/valhalla/issues/2830#issuecomment-772468377

1. Fix when libprotobuf is of a lower version
2. Fix when boost is of a lower version.

### CODE REVIEW

File: src/mjolnir/valhalla_add_predicted_traffic.cc [hunk 1]
```diff
@@ -173,7 +173,7 @@ void update_tile(const std::string& tile_dir,
   size_t pred_count = 0;
   for (uint32_t j = 0; j < tile_builder.header()->directededgecount(); ++j) {
     auto found = speeds.find(j);
-    pred_count += found != speeds.cend() && found->second.coefficients.has_value();
+    pred_count += found != speeds.cend() && bool(found->second.coefficients);
```
Comment (xlqian) [inline src/mjolnir/valhalla_add_predicted_traffic.cc:176]: https://www.boost.org/doc/libs/1_68_0/libs/optional/doc/html/boost_optional/reference/header__boost_optional_optional_hpp_/detailed_semantics___optional_values.html

basically, `bool()` and `has_value` is the same thing

Comment (kevinkreiser) [reply src/mjolnir/valhalla_add_predicted_traffic.cc:176]: yep i noticed this too. would you mind using `static_cast<bool>()` for these?

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 * **Removed**
 
 * **Bug Fix**
+   * FIXED: Fix compilation errors when boost < 1.68 and libprotobuf < 3.6  [#2878](https://github.com/valhalla/valhalla/pull/2878)
```
Comment (kevinkreiser) [inline CHANGELOG.md:5]: can you move this to the bottom of the list, we've been following that pattern. basically inside of a release the oldest stuff is at the top and the youngest stuff is at the bottom. i can see why that is somewhat unappealing though.... maybe we can change it in the next release and follow that from now on

File: src/mjolnir/valhalla_add_predicted_traffic.cc (hunk 1 above)
Comment (xlqian) [reply src/mjolnir/valhalla_add_predicted_traffic.cc]: yeah sure, but for me, `bool()` is a defined specialized  explicit operator from `boost::optional`(even in lower version) class... it's not a "just" casting :)

File: src/thor/route_matcher.cc [hunk 3]
```diff
@@ -191,7 +191,7 @@ bool expand_from_node(const mode_costing_t& mode_costing,
         elapsed += cost;
         // overwrite time with timestamps
         if (use_timestamps)
-          elapsed.secs = shape[index].time() - shape[0].time();
+          elapsed.secs = shape.Get(index).time() - shape.Get(0).time();
```
Comment (xlqian) [inline src/thor/route_matcher.cc:194]: hmmm, how can I test this so that codecov would be contented?

Comment (kevinkreiser) [reply src/thor/route_matcher.cc:194]: that shouldnt be your burden. we can merge and just ignore codecov in this instance. this is for the fast exact map matching endpoint which as a whole could use more structured testing.

**Review by kevinkreiser — APPROVED** [review]

**Pull request merged**